### PR TITLE
fix DEX format

### PR DESCRIPTION
**Pull request by h4ng3r**
+ enhance DEX format parsing
+ fix problem with UTF-8 methods, fields and classes
+ add signature to all methods and fields (in dex they can have the same name)
+ fix checksum check
+ delete some old and unused code

### CODE REVIEW

**Comment by h4ng3r** [comment]
The test has been modified: https://github.com/radare/radare2-regressions/pull/538

File: libr/bin/format/dex/dex.c [hunk 1]
```diff
@@ -144,7 +175,26 @@ RBinDexObj *r_bin_dex_new_buf(RBuffer *buf) {
 		bin->fields[i].type_id = r_read_le16 (bufptr + offset + 2);
 		bin->fields[i].name_id = r_read_le32 (bufptr + offset + 4);
 	}
+
+	/* proto */
+    int protos_size = dexhdr->prototypes_size * sizeof (struct dex_proto_t);
```
Comment (radare) [inline libr/bin/format/dex/dex.c]: use tabs

File: libr/bin/p/bin_dex.c [hunk 2]
```diff
@@ -18,6 +18,207 @@
...
+	if (idx < 0 || idx >= bin->header.strings_size || !bin->strings) {
+		return NULL;
+	}
+
+	r_buf_read_at (bin->b, bin->strings[idx], buf, sizeof (buf));
+	uleblen = r_uleb128 (buf, sizeof (buf), &len) - buf;
+
+	// TODO: improve this ugly fix
+	char c = "a";
+	while(c != 0) {
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: missing space. unnecesessary !=0

File: libr/bin/p/bin_dex.c [hunk 3]
```diff
@@ -18,6 +18,207 @@
...
+	int uleblen;
+	if (idx < 0 || idx >= bin->header.strings_size || !bin->strings) {
+		return NULL;
+	}
+
+	r_buf_read_at (bin->b, bin->strings[idx], buf, sizeof (buf));
+	uleblen = r_uleb128 (buf, sizeof (buf), &len) - buf;
+
+	// TODO: improve this ugly fix
+	char c = "a";
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: "" is not char you should use ''

File: libr/bin/p/bin_dex.c [hunk 4]
```diff
@@ -18,6 +18,207 @@
...
+    count = (((val + (val >> 4)) & 0x0F0F0F0F) * 0x01010101) >> 24;
+
+    return count;
+}
+
+/*
+ * Flag for use with createAccessFlagStr().
+ */
+typedef enum {
+    kAccessForClass = 0, kAccessForMethod = 1, kAccessForField = 2,
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: use tabs

File: libr/bin/p/bin_dex.c [hunk 5]
```diff
@@ -73,27 +264,28 @@ static int check_bytes(const ut8 *buf, ut64 length) {
...
 	// M5 (Feb-Mar 08)
-        if (!memcmp (buf, "dex\n009\0", 8)) {
-	        return true;
+		if (!memcmp (buf, "dex\n009\0", 8)) {
+			return true;
 	}
 	// Default fall through, should still be a dex file
 	if (!memcmp (buf, "dex\n", 4)) {
-                return true;
+				return true;
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: wtf

File: libr/bin/p/bin_dex.c [hunk 6]
```diff
@@ -73,27 +264,28 @@ static int check_bytes(const ut8 *buf, ut64 length) {
...
 	}
 	return false;
 }
 
+
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: drop that line

File: libr/bin/p/bin_dex.c [hunk 7]
```diff
@@ -589,14 +721,19 @@ static int *parse_class(RBinFile *binfile, RBinDexObj *bin, RBinDexClass *c, RBi
...
-	dprintf ("  virtual methods: %u\n", (ut32)VM);
+	dprintf ("  Virtual methods   -\n");
+    omi = 0;
 	for (i = 0; i < VM; i++) {
 		ut64 MI, MA, MC;
 		p = r_uleb128 (p, p_end-p, &MI);
 		p = r_uleb128 (p, p_end-p, &MA);
 		p = r_uleb128 (p, p_end-p, &MC);
 
+        MI += omi;
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: wrongly indented. use tabs

File: libr/bin/p/bin_dex.c [hunk 8]
```diff
@@ -589,14 +721,19 @@ static int *parse_class(RBinFile *binfile, RBinDexObj *bin, RBinDexClass *c, RBi
 		}
 		free (method_name);
 	}
+
 	/* virtual methods */
-	dprintf ("  virtual methods: %u\n", (ut32)VM);
+	dprintf ("  Virtual methods   -\n");
+    omi = 0;
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: use tabs

File: libr/bin/p/bin_dex.c [hunk 9]
```diff
@@ -657,49 +802,43 @@ static int dex_loadcode(RBinFile *arch, RBinDexObj *bin) {
 		return false;
 	}
 
-	dprintf ("Walking %d classes\n", bin->header.class_size);
+	/* debug prototypes */
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: use tabs

**Pull request merged**